records/2021-11-30.performance-testing.md

```diff
@@ -0,0 +1,127 @@
+
+AWS Lambda functions allow code to be run in a highly scalable way. They can integrate properly with AWS products and in this case we are required to run the performance tests once every day after which the results (HTML and CSV) are published to S3. This option was explored because: 
+
+- It has less complexity for the Notify team to manage, the code was lesser compared to ECS.
+- Costs lesser than ECS
+
+We initially went with this option, but soon found out that the file descriptors were capped and the performance tests were quickly running out of connections
+
+> POST,/v2/notifications/email,
+"HTTPSConnectionPool(host='api.staging.notification.cdssandbox.xyz', port=443): 
+Max retries exceeded with url: /v2/notifications/email 
+(Caused by SSLError(OSError(24, 'No file descriptors available')))",1
+
+
+#### Elastic Container Service(ECS)
+
+Given the aforementioned experience with Lambda, we turned to ECS mainly because of the limited file descriptors available.
+
+Amazon Elastic Container Service (Amazon ECS) offers high scalability, faster container management service, stop, and manage containers. Our Performance test containers are defined in a task definition that used to execute the performance test a service. The service configuration enables the performance test to run and maintain a pre-defined number of tasks. The tasks and services are on a serverless infrastructure that is managed by AWS Fargate [[3]](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/Welcome.html). 
+
+Amazon ECS enables you to launch and stop our performance test container by using schedule expressionm defined in the task [[3]](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/Welcome.html).
+
+The scheduled placement of performance test container are based on resource needs, isolation policies, and availability requirements [[3]](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/Welcome.html). 
+
+Following the successful execution of the performance test, HTML and CSV results are published to a staging S3 bucket. 
+
+An hour after the ECS has finished the performance test, the results are exported to Github via this workflow script [[4]](https://github.com/cds-snc/notification-performance-test-results/blob/main/.github/workflows/sync-performance-test-results.yml
+).
+
+
+![Overview of the performance test container ECS agent](diagrams/2021-11-30.performance-testing/overview-notify-performance-test-containeragent.png?raw=true "Overview of the performance test container ECS agent")
+
+## Decision
+
+The decision was to go with:
+
+-  Locust as it is the most compatible with the project and team. Being that the tool is Python and open source, it can be exploited to be the most suitable to performance test tool for testing throughput of the Notify application.
+- Though Lambda is widely used and easier, we decided to go with the container agent solution from AWS ECS.
```

Comment:
I would add the reason here as it is a summary - because of the file descriptor issue.

Reply:
Ok will update